### PR TITLE
Restrict ppx_deriving_yojson dependency to yojson < 1.6

### PR DESCRIPTION
**Pull request by NathanReb**
Yojson 1.6 deprecates the type `json` in favor of the new type `t`.
This shouldn't be an issue but users of ppx_deriving_yojson and
dune are gonna have a hard time working with yojson 1.6 since warning 3
is fatal in the default dune profile and they can't change
ppx_deriving_yojson output.

### CODE REVIEW

**Comment by NathanReb** [comment]
Looks like the usual revdeps failures, let me know if I missed one that I actually broke!

**Comment by NathanReb** [comment]
Pushed some changes to hopefully make the linter happy and fix the broken builds for the older versions of `ppx_deriving_yojson`.

File: packages/ppx_deriving_yojson/ppx_deriving_yojson.1.0/opam [hunk 1]
```diff
@@ -4,8 +4,8 @@ authors: [ "Peter Zotov <whitequark@whitequark.org>" ]
 license: "MIT"
 homepage: "https://github.com/whitequark/ppx_deriving_yojson"
 doc: "http://whitequark.github.io/ppx_deriving_yojson"
-#bug-reports: "https://github.com/whitequark/ppx_deriving_yojson/issues"
-#dev-repo: "git://github.com/whitequark/ppx_deriving_yojson.git"
+bug-reports: "https://github.com/whitequark/ppx_deriving_yojson/issues"
+dev-repo: "git://github.com/whitequark/ppx_deriving_yojson.git"
```
Comment (avsm) [inline packages/ppx_deriving_yojson/ppx_deriving_yojson.1.0/opam]:
```suggestion
# dev-repo: "git://github.com/whitequark/ppx_deriving_yojson.git"
```

the line is duplicated below

File: packages/ppx_deriving_yojson/ppx_deriving_yojson.1.1/opam [hunk 2]
```diff
@@ -4,8 +4,8 @@ authors: [ "Peter Zotov <whitequark@whitequark.org>" ]
 license: "MIT"
 homepage: "https://github.com/whitequark/ppx_deriving_yojson"
 doc: "http://whitequark.github.io/ppx_deriving_yojson"
-#bug-reports: "https://github.com/whitequark/ppx_deriving_yojson/issues"
-#dev-repo: "git://github.com/whitequark/ppx_deriving_yojson.git"
+bug-reports: "https://github.com/whitequark/ppx_deriving_yojson/issues"
+dev-repo: "git://github.com/whitequark/ppx_deriving_yojson.git"
```
Comment (avsm) [inline packages/ppx_deriving_yojson/ppx_deriving_yojson.1.1/opam]:
```suggestion
# dev-repo: "git://github.com/whitequark/ppx_deriving_yojson.git"
```

field is duplicated below

**Comment by NathanReb** [comment]
Ouh my bad, missed the camelus notification on Friday!

**Comment by NathanReb** [comment]
Hmm now there's an issue with ppx_import, I'll look into it!

**Comment by mseri** [comment]
Looks like the only remaining issue is with `ppx_import` for `ppx_deriving_yojson.1.0` on `ocaml-4.03.0`.

The rest:
- `nsq` seems to have been fixed yesterday with https://github.com/ocaml/opam-repository/pull/13394
- some old versions of `ocaml-jupyter` have issues with `omp`, probably a missing upper bound or just plain incompatibility with newer compilers
- `links` is not compatible with newer compilers

**Comment by kit-ty-kate** [comment]
Everything seems fine except on 4.02.3:
```
#=== ERROR while compiling ppx_deriving_yojson.1.0 ============================#
# context     2.0.0 | linux/x86_64 | ocaml-base-compiler.4.02.3 | file:///home/travis/build/ocaml/opam-repository
# path        ~/.opam/ocaml-base-compiler.4.02.3/.opam-switch/build/ppx_deriving_yojson.1.0
# command     ~/.opam/opam-init/hooks/sandbox.sh build ocamlbuild -classic-display -use-ocamlfind src_test/test_ppx_yojson.byte --
# exit-code   10
# env-file    ~/.opam/log/ppx_deriving_yojson-15051-03b1a4.env
# output-file ~/.opam/log/ppx_deriving_yojson-15051-03b1a4.out
### output ###
# ocamlfind ocamldep -package oUnit -package yojson -modules src_test/test_ppx_yojson.ml > src_test/test_ppx_yojson.ml.depends
# ocamlfind ocamlc -c -g -bin-annot -safe-string -ppx 'ocamlfind ppx_import/ppx_import' -ppx 'ocamlfind ppx_deriving/ppx_deriving src/ppx_deriving_yojson.cma' -package oUnit -package yojson -w @5@8@10@11@12@14@23@24@26@29@40 -I src_test -I src -o src_test/test_ppx_yojson.cmo src_test/test_ppx_yojson.ml
# + ocamlfind ocamlc -c -g -bin-annot -safe-string -ppx 'ocamlfind ppx_import/ppx_import' -ppx 'ocamlfind ppx_deriving/ppx_deriving src/ppx_deriving_yojson.cma' -package oUnit -package yojson -w @5@8@10@11@12@14@23@24@26@29@40 -I src_test -I src -o src_test/test_ppx_yojson.cmo src_test/test_ppx_yojson.ml
# ocamlfind: Package `ppx_import' not found
# File "src_test/test_ppx_yojson.ml", line 1:
# Error: Error while running external preprocessor
# Command line: ocamlfind ppx_import/ppx_import '/tmp/camlppxf2b7c6' '/tmp/camlppx741f78'
# 
# Command exited with code 2.
```
I'm not sure how this change introduces a failure as it seems that it was working in 3ae7241ca8af9843b16eb8940c8be638c76da856: http://check.ocamllabs.io/4.02.3/good/ppx_deriving_yojson.1.0

**Comment by kit-ty-kate** [comment]
ah, got it.. this package tests during normal compilation

**Comment by NathanReb** [comment]
Ouh sorry, got carried away, will have a look at it, thanks for the help!

**Comment by NathanReb** [comment]
I added a missing test dependency in the 1.1 package but that won't solve the issue. At this point I don't see what I could do beside removing the test build since the tests appear to be broken.

**Comment by NathanReb** [comment]
I tried running `opam install -t` for various versions of `ppx_deriving_yojson` on a local `4.02.3` switch and several of them are somehow broken so this is indeed unrelated to this PR.

I'll disable the test builds for the broken versions.

**Comment by NathanReb** [comment]
There's still an issue with 3.x which tests won't build on 4.02 for some reason but will for later versions of ocaml.

Not sure what the best solution is here, maybe add an `ocaml >= 4.03` lower bound on `3.0`.

Considering that the package was already broken before this PR it might be worth considering merging it. I'm happy to fix those old versions but not at the expense of the important part (ie the upper bound on `yojson`).

**Comment by mseri** [comment]
If it does not compile, why not adding `ocaml >= 4.03`?
Then I'd be happy to merge. Thanks for the help

**Comment by NathanReb** [comment]
It does compile, the tests don't ie `opam install` will work but `opam install -t` won't. That's why the CI fails from what I understood.

**Comment by kit-ty-kate** [comment]
it looks like this works. I'll let the CI complete but it seems good

**Comment by NathanReb** [comment]
I'd expect the 4.02.3 build to still fail, just on a later version, 2.x ones have being fixed indeed. If I'm wrong and it actually goes through then that's even better !

**Comment by kit-ty-kate** [comment]
> I'd expect the 4.02.3 build to still fail, just on a later version, 2.x ones have being fixed indeed. If I'm wrong and it actually goes through then that's even better !

well no, the build is already finished and everything is fine on 4.02.3 ^^

**Comment by kit-ty-kate** [comment]
all green. Thanks a lot

**Pull request merged**